hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/BlockCacheFactory.java

```diff
@@ -113,7 +117,17 @@ public static BlockCache createBlockCache(Configuration conf) {
         LOG.warn(
           "From HBase 2.0 onwards only combined mode of LRU cache and bucket cache is available");
       }
-      return bucketCache == null ? l1Cache : new CombinedBlockCache(l1Cache, bucketCache);
+
+      if (bucketCache == null) {
+        return l1Cache;
+      }
+
+      if (conf.getBoolean(BLOCKCACHE_VICTIM_HANDLER_ENABLED_KEY,
+          BLOCKCACHE_VICTIM_HANDLER_ENABLED_DEFAULT)) {
+        return new InclusiveCombinedBlockCache(l1Cache, bucketCache);
```

Comment:
This mode was removed in 2.x line.  Reasoning to add it back?

Reply:
Bucket cache is no longer as a victim handler for LRU by [HBASE-19357](https://issues.apache.org/jira/browse/HBASE-19357)

In this situation, the L1 LRU cache's memory is totally on heap.

If the index and bloom filters on a region server are big, then a big LRU cache will introduce more GC cost.

So we should add a configuration for user to choose victim handler.

A small LRU cache with a big victim offheap bucket cache can reduce GC cost.

Reply:
The bucket cache may waste a little memory space, and the serialization/deserialization may introduce a little more cpu cost/time, but the GC pause time is reduced significantly on big RAM machine with big index and bloomfilters. And machines with big RAM such as 256GB or 320GB or even 448GB are more and more common.

This is also a configuration, users can decide to enable it or not, based on their priority.

Reply:
yes, bucket cache is slower.

Reply:
The call queue time  and the call process time are also improved.

<img width="1127" alt="WeChatWorkScreenshot_c67d3379-5f00-4628-b635-2bd42f9ad68f" src="https://user-images.githubusercontent.com/4401756/169735032-47bec0bc-51cc-4083-98f2-490641dec41d.png">


<img width="1056" alt="WeChatWorkScreenshot_7d9b0937-8464-4abf-a231-f1ebf0f66700" src="https://user-images.githubusercontent.com/4401756/169736345-d4bcb488-5900-44cf-a5a3-8dc637049594.png">

Reply:
This change makes sense to me. I was disappointed to see the victim handling removed in 2.x, and I think it would be good to add it back. 

We have some regionservers where the index + bloom is too big to fit into an on-heap cache. Maybe BucketCache is slower, but it's still better than the high eviction rate when your on-heap cache can't fit all of the L1 objects. 

Is there is a real reason to block this?

Reply:
@liangxs what do you think about having the victim cache still be preferential?

DATA goes direct to BucketCache
META goes direct to LRU, and victims flush to BucketCache

This seems like a good compromise of the performance. META still prefers faster LRU on-heap, but if too big to fit ends up in BucketCache.

Reply:
@liangxs are the above graphs using master branch or branch-2?

Reply:
using branch-2.4

Reply:
Ok that makes sense. I've been looking at the implementation of the `victimHandler` code in LruBlockCache. A cache miss in L1 which is found in L2 [is promoted back to L1](https://github.com/apache/hbase/blob/branch-2/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/LruBlockCache.java#L512-L517). It seems like that would cause your change to not really improve GC, because we'd continually have META blocks going back and forth between L1 and L2.

So your results were confusing to me, and I also tested this and my results were similar to yours. But then I noticed in branch-2.4 (and 2), we don't call [`L1.getBlock` unless `L1.containsBlock`](https://github.com/apache/hbase/blob/branch-2/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CombinedBlockCache.java#L81). This circumvents the victimHandler code above.

But here in master branch, an [improvement was made](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CombinedBlockCache.java#L85-L96) to remove the unnecessary containsBlock call when we know the BlockType (most cases). So I think in master branch (or if that improvement is ever ported to branch-2) we'd probably stop seeing the GC improvement.

So I was thinking about this and without the victim promotion back to L1 we basically have a case where certain blocks will get into L2 and never get back to L1. This artificially benefits META blocks which may be read later in the lifetime of the regionserver, because there is less competition for L1 space since more blocks will have flushed to L2. This doesn't quite feel like a real L1/L2 cache setup, instead it feels like a single cache where some blocks randomly exist on-heap and some off-heap. 

So I do agree with you that we need to have a solution for the case where META blocks are much larger than the possible L1 size, but the current approach feels like an accidental solution rather than a complete one. A few thoughts I had:

1. Just disable LRU, use a single unified BucketCache as the L1. Personally I don't think the BucketCache performs that much worse than LRU. In my tests it's always an improvement over LRU, and @anoopsjohn's reports from years ago Alibaba tests show that BucketCache is 30% faster. 
2. Use LruAdaptiveBlockCache as the L1, which would help reduce the amount of churn between the L1 and L2. This algorithm is cool, but comes with many complicated configurations.
3. Enhance the victim handler code to reduce the likelihood of promoting a victim back into L1. A good example here might be to inspect the accessCount on the victim block and only promote after multiple access.

I'm guessing option 3 might be the most generally acceptable solution, if it's possible. I think the simplest form would check the BlockPriority of the L2-cached block and only promote when it's BlockPriority.MULTI.